packages/compiler/core/checker.ts

```diff
@@ -4674,33 +4671,54 @@ export function filterModelProperties(
   return finishTypeForProgram(program, newModel);
 }
 
+/**
+ * Gets the property from the nearest base type that is overridden by the
+ * given property, if any.
+ */
+export function getOverriddenProperty(property: ModelProperty): ModelProperty | undefined {
+  compilerAssert(
+    property.model,
+    "Parent model must be set before overridden property can be found."
+  );
+
+  for (let current = property.model.baseModel; current; current = current.baseModel) {
+    const overriden = current.properties.get(property.name);
+    if (overriden) {
+      return overriden;
+    }
+  }
+
+  return undefined;
+}
+
+/**
+ * Enumerates the properties declared by model or inherited from its base.
+ *
+ * Properties declared by more derived types are enumerated before properties
+ * of less derived types.
+ *
+ * Properties that are overridden are not enumerated.
+ */
 export function* walkPropertiesInherited(model: Model) {
-  let current: Model | undefined = model;
-  const allReturnedProps = new Map<string, ModelProperty>();
-  while (current) {
-    const returnProps = Array<ModelProperty>();
-    for (const prop of current.properties.values()) {
-      if (!allReturnedProps.has(prop.name)) {
-        allReturnedProps.set(prop.name, prop);
-        returnProps.push(prop);
+  const returned = new Set<string>();
+
+  for (let current: Model | undefined = model; current; current = current.baseModel) {
+    for (const property of current.properties.values()) {
+      if (returned.has(property.name)) {
+        // skip properties that have been overridden
+        continue;
       }
+      returned.add(property.name);
+      yield property;
     }
-    yield* returnProps;
-    current = current.baseModel;
```

Comment:
Cleaned this up a bit to use the same pattern for looping through bases and avoid making intermediate arrays.